Automattic/wp-calypso
# Framework: single tree rendering

Single tree render React element tree.

- [The GH project](https://github.com/Automattic/wp-calypso/projects/53)
- [Codemod](https://github.com/Automattic/wp-calypso/blob/master/bin/codemods/src/single-tree-rendering.js) that does the majority of the work here.

# Testing
Smoke test every route and every view:

- Anything that changes URL is interesting; there are no changes to UI elements like dropdowns etc.
- I’m especially curious about how Woocommerce store sections work (since those parts are built quite differently than rest of the Calypso) and domains (just tricky to test).
- Ensure all routes work. If something gets stuck, it’s probably just because of misplaced `next()` in middleware.
- Ensure sidebar is hidden (or visible) when it should.
- Ensure routes under “My sites” work when the user doesn’t have any sites, when the user has Jetpack site selected as well when the user has simple wordpress.com site selected.
- Look for warnings @ console.

# What has changed

### Previously

Calypso renders content separately in `#primary` (main content) and `#secondary` (sidebar) elements but this has many caveats.

Defining route:
```js
page( '/hello-dolly', navigation, helloDolly );
```

Controller:
```js
const navigation = ( context, next ) => {
  renderWithReduxStore(
    <Navigation />,
    document.getElementById( 'secondary' ),
    context.store
  );
  next();
};

const helloDolly = context => {
  renderWithReduxStore(
    <HelloDolly />,
    document.getElementById( 'primary' ),
    context.store
  );
};
```

### In future

Instead of rendering in multiple elements using `ReactDom.render()`, we want to render everything under single root element, i.e. a single tree.

Defining route:
```js
import { makeLayout, render as clientRender } from 'controller';

page( '/hello-dolly', navigation, helloDolly, makeLayout, clientRender );
```

Controller:
```js
const navigation = ( context, next ) => {
  context.secondary = <Navigation />;
  next();
};

const helloDolly = ( context, next ) => {
  context.primary = <HelloDollyPage />;
  next();
};
```

### What are `makeLayout` & `clientRender`?

`makeLayout` middleware builds one element from `primary` and `secondary` properties.

Little simplified but:
```js
context.layout = (
  <ReduxProvider store={ store }>
    <Layout
      primary={ primary }
      secondary={ secondary }
    />
  </ReduxProvider>
);
```

`clientRender` middleware does the final `ReactDom.render()` — that should be the only one we have anywhere.

```js
ReactDom.render(
  context.layout,
  document.getElementById( 'wpcom' )
);
```

### Don't unmount

We don't have to unmount elements anymore because everything — including the sidebar — is under single element tree and React handles reconciliation on changes. Previously sidebar would be hidden like so:

```js
ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );
```

In future, you simply leave `context.secondary` undefined or set it `null`.

### Don't wrap in Redux provider anymore
You don't need to wrap your components with `<ReduxProvider />` anymore since our only element tree is already wrapped with it. See `makeLayout` middleware above.

# How to re-run this codemod
This is just note to myself. If you're just testing this PR, you don't need to care about this.

1. Run prettier to confirm what files would be modified by it later, so that we can ignore them:
    ```bash
    npm run reformat-files
    ```
    Copy the list, you'll need it later, then reset modifications:
     ```bash
    git checkout HEAD -- .
    ```

1. Run the codemod
    ```bash
    npm run codemod single-tree-rendering client
    ```

1. Ignore files we don't want to codemod
    ```bash
    git checkout HEAD -- \
     ./client/boot/common.js \
     ./client/controller/index.web.js \
     ./client/boot/project/wordpress-com.js \
     ./client/lib/react-helpers/index.js \
     ./client/components/tinymce/plugins/mentions/plugin.jsx \
     ./client/layout/guided-tours/tours/main-tour.js \
     ./client/lib/accept/index.js \
     ./client/blocks/site/index.jsx
    ```

1. Run Prettier
    ```bash
    npm run reformat-files
    ```

1. Ignore files we don't want to prettify (check if your list is still up to date from the 1st step)
    ```bash
    git checkout HEAD -- \
      ./client/blocks/login/two-factor-authentication/verification-code-form.jsx \
      ./client/components/data/query-posts/index.jsx \
      ./client/components/domains/registrant-extra-info/fr-form.jsx \
      ./client/components/web-preview/toolbar.jsx \
      ./client/extensions/zoninator/components/forms/zone-content-form/post-card.jsx \
      ./client/extensions/zoninator/components/forms/zone-content-form/posts-list.jsx \
      ./client/layout/masterbar/drafts.jsx \
      ./client/login/magic-login/emailed-login-link-successfully.jsx \
      ./client/state/analytics/actions.js \
      ./client/state/analytics/test/actions.js \
      ./client/state/posts/reducer.js \
      ./client/state/posts/test/reducer.js
    ```

1. Fix these errors:
    ```bash
     client/my-sites/controller.js
       154:17  error  'reduxStore' is assigned a value but never used  no-unused-vars
     
     client/my-sites/sharing/controller.js
       28:34  error  'store' is assigned a value but never used  no-unused-vars
    ```

1. Commit.
